### PR TITLE
chore: data pipeline tests improvements

### PR DESCRIPTION
**Pull request by mrehan27**
helps: https://linear.app/customerio/issue/MOB-12/improve-cdp-test-coverage

###  Changes

- Separated utility extension methods and plugins
- Improved null checks and error messages in tests
- Added extension to make dictionary comparison easier in tests
- Fixed broken methods

### CODE REVIEW

File: Tests/DataPipeline/Support/Segment.swift [hunk 1]
```diff
@@ -75,27 +77,14 @@ extension RawEvent {
...
-func waitUntilStarted(analytics: Analytics?) {
-    guard let analytics = analytics else { return }
-
-    // wait until the startup queue has emptied it's events.
-    if let startupQueue = analytics.find(pluginType: StartupQueue.self) {
-        while startupQueue.running != true {
-            RunLoop.main.run(until: Date.distantPast)
+extension Analytics {
+    /// Attaches and returns plugin only if it wasn't attached previously
+    func addPluginOnce<P: Plugin>(plugin: P) -> P {
```
Comment (mrehan27) [inline Tests/DataPipeline/Support/Segment.swift:82]: Renamed for clarity

File: Tests/DataPipeline/Support/XCTestCaseExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import Foundation
+import XCTest
+
+extension XCTest {
+    /// Asserts that the actual dictionary matches the expected dictionary, with type-specific comparisons. Assumes `String` as the default type for values unless explicitly specified.
+    func XCTAssertMatches(
```
Comment (mrehan27) [inline Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: This method has helped me reduce boilerplate code in dictionary comparisons

Comment (levibostian) [reply Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: Could you give some examples of code that uses this, showing the reason why it's needed? 

I feel like something like this should already exist in the XCTest framework? So I wanted to ask this question to see. I would prefer we can leverage something already in the framework then maintain this function in the long-term.

Comment (levibostian) [reply Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: If we do decide to keep this, I suggest moving it into `SharedTest` module so all test modules can use it.

File: Tests/DataPipeline/Support/Segment.swift (hunk 1 above)
Comment (mrehan27) [reply Tests/DataPipeline/Support/Segment.swift]: Removed this completely for simplicity

File: Tests/DataPipeline/Support/XCTestCaseExtensions.swift (hunk 2 above)
Comment (mrehan27) [reply Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: > Could you give some examples of code that uses this

I've utilized it for checking user traits and other event properties. You can its usage in [`DataPipelineInteractionTests`](https://github.com/customerio/customerio-ios/blob/rehan/cdp-test-base/Tests/DataPipeline/DataPipelineInteractionTests.swift) and [`DataPipelineCompatibilityTests`](https://github.com/customerio/customerio-ios/blob/rehan/cdp-json-compat-tests/Tests/DataPipeline/DataPipelineCompatibilityTests.swift).

> I feel like something like this should already exist in the XCTest framework

I couldn't find a similar method in the XCTest framework. It may be likely due to `Any` not being `Equatable`, making direct comparison of `[String : Any]` challenging. I've written this method to so we can transform `Any` into some `Equatable` and compare.

> I suggest moving it into SharedTest module so all test modules can use it

I decided not to include it in `SharedTest` as its current use seemed limited outside the `DataPipeline` module. But I'm open to relocating it if it's more beneficial or easier to maintain there.

File: Sources/DataPipeline/DataPipeline.swift [hunk 3]
```diff
@@ -32,9 +32,9 @@ public class DataPipeline: ModuleTopLevelObject<DataPipelineInstance>, DataPipel
     #if DEBUG
     /// Updates the implementation instance. To be used for testing purposes only.
     static func createAndSetSharedTestInstance(diGraphShared: DIGraphShared, config: DataPipelineConfigOptions) -> DataPipelineImplementation {
+        moduleConfig = config
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Sources/DataPipeline/DataPipeline.swift:36]: Added comment.

File: Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift [hunk 4]
```diff
@@ -14,17 +14,25 @@ class DataPipelineImplementationScreenViewsTest: IntegrationTest {
     override func setUp() {
         super.setUp()
 
-        let analytics = CustomerIO.shared.analytics
+        guard let analytics = CustomerIO.shared.analytics else {
+            fatalError("Analytics instance is nil. The SDK has been set up incorrectly.")
+        }
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift]: I like the idea, I've implemented it and mode more changes to simplify the implementation in `DataPipeline` tests.

File: Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift [hunk 5]
```diff
@@ -14,17 +14,25 @@ class DataPipelineImplementationScreenViewsTest: IntegrationTest {
...
+
         // setting up required plugins
-        outputReader = attachPlugin(analytics: analytics, plugin: OutputReaderPlugin())
-        autoTrackingScreenViews = attachPlugin(analytics: analytics, plugin: AutoTrackingScreenViews())
+        outputReader = analytics.addPluginOnce(plugin: OutputReaderPlugin())
+        autoTrackingScreenViews = analytics.addPluginOnce(plugin: AutoTrackingScreenViews())
         // wait for analytics queue to start emitting events
-        waitUntilStarted(analytics: analytics)
+        analytics.waitUntilStarted()
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift]: This isn't really common only for plugins, as it mainly ensures `analytics` starts emitting events before we start verifying test values.

Following the [suggestion above](https://github.com/customerio/customerio-ios/pull/484#discussion_r1464557241), I've moved `waitUntilStarted` in base test classes for `DataPipeline` module.

File: Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift [hunk 6]
```diff
@@ -14,17 +14,25 @@ class DataPipelineImplementationScreenViewsTest: IntegrationTest {
...
 
         // Screenview events are ignored if no profile identified
         CustomerIO.shared.identify(identifier: String.random)
     }
 
+    override func tearDown() {
+        CustomerIO.shared.clearIdentify()
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift:23]: The main reason for calling `clearIdentify` is to reset `analytics` instance and its storage. I agree base `tearDown` could do more, but then this can raise more challenges impacting other tests which currently I didn't plan for this PR and the ticket.

For now, I think it is okay to keep here. But gradually, update the base test to do more and help remove boilerplate code from subclasses so it is consistent and easier to maintain.

File: Tests/DataPipeline/DataPipelineInteractionTests.swift [hunk 7]
```diff
@@ -31,10 +31,13 @@ class DataPipelineInteractionTests: UnitTest {
...
         analytics = customerIO.analytics
+        guard let analytics = customerIO.analytics else {
+            fatalError("Analytics instance is nil. The SDK has been set up incorrectly.")
+        }
         // OutputReaderPlugin helps validating interactions with analytics
-        outputReader = attachPlugin(analytics: analytics, plugin: OutputReaderPlugin())
+        outputReader = analytics.addPluginOnce(plugin: OutputReaderPlugin())
         // wait for analytics queue to start emitting events
-        waitUntilStarted(analytics: analytics)
+        analytics.waitUntilStarted()
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift]: Updated

File: Tests/DataPipeline/DataPipelineInteractionTests.swift [hunk 8]
```diff
@@ -58,35 +61,33 @@ class DataPipelineInteractionTests: UnitTest {
         XCTAssertEqual(analytics.traits()?.count, 0)
 
         XCTAssertEqual(outputReader.identifyEvents.count, 1)
+        guard let identifyEvent = outputReader.identifyEvents.last else {
+            XCTFail("captured event must not be nil")
+            return
+        }
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift:50]: Unfortunately we can't use force call `!` here. As for cases where number of events is not equal to 1 or event casting fails, the guard statement will prevent crashing tests and fail it gracefully. While adding a force call will crash the test because event may be `nil`. We could use optional calls `?` only, but I found `guard` statement more clear and safer to prevent false positives when writing tests.

File: Tests/DataPipeline/DataPipelineInteractionTests.swift [hunk 9]
```diff
@@ -58,35 +61,33 @@ class DataPipelineInteractionTests: UnitTest {
...
 
-        let identifyEvent = outputReader.identifyEvents.first
-        XCTAssertEqual(identifyEvent?.userId, givenIdentifier)
-        XCTAssertEqual(identifyEvent?.traits?.dictionaryValue?.count, 0)
+        XCTAssertEqual(identifyEvent.userId, givenIdentifier)
+        XCTAssertEqual(identifyEvent.traits?.dictionaryValue?.count, 0)
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift]: I agree with the test assumption here. And there are a lot of objects that Segment attaches to an event, and they are fine to have. But the properties for `identify` event go to `traits`, and if the count is 0 here, this confirms that there are no other properties attached to `identify` event.

For other areas, can you please point out a few so I can take a look at them too?

PS: I updated count check to match with empty dict for consistency with rest of the code.

File: Tests/DataPipeline/DataPipelineInteractionTests.swift [hunk 10]
```diff
@@ -58,35 +61,33 @@ class DataPipelineInteractionTests: UnitTest {
...
     }
 
     func test_identify_expectSetNewProfileWithAttributes() {
         let givenIdentifier = String.random
         let givenBody: [String: Any] = ["first_name": "Dana", "age": 30]
+        let traitsTypeMap: [[String]: Any.Type] = [["age"]: Int.self]
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift]: Renamed to `givenBodyTypeMap` for clarity and consistency with `givenBody`

File: Tests/DataPipeline/Support/MockExtensions.swift [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            attributes["device_manufacturer"] = deviceManufacturer
+        }
+        if let deviceModel = underlyingDeviceModel {
+            attributes["device_model"] = deviceModel
+        }
+        if let deviceOsVersion = underlyingOsVersion {
+            attributes["device_os"] = deviceOsVersion
+        }
+        attributes["push_enabled"] = String(isPushSubscribed)
+        return attributes
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/Support/MockExtensions.swift]: Yes, it surely does, I missed it before. Thanks for sharing.

File: Tests/DataPipeline/Support/Segment.swift [hunk 12]
```diff
@@ -75,27 +77,14 @@ extension RawEvent {
...
+        if find(pluginType: P.self) != nil {
+            fatalError("Plugin \(P.self) is already attached")
```
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/Support/Segment.swift]: I simplified it by using default `add` method to avoid any ambiguities 😄

File: Tests/DataPipeline/UnitTest.swift [hunk 13]
```diff
@@ -0,0 +1,32 @@
+@testable import CioDataPipelines
+@testable import CioInternalCommon
+import Foundation
+@testable import Segment
+import SharedTests
+
+class UnitTest: SharedTests.UnitTest {
+    // When calling CustomerIOInstance functions in the test functions, use this `CustomerIO` instance.
+    // This is a workaround until this code base contains implementation tests. There have been bugs
```
Comment (Shahroz16) [inline Tests/DataPipeline/UnitTest.swift:9]: I don't quite understand this, and is it still needed?

**Review by Shahroz16 — APPROVED** [review]

File: Tests/DataPipeline/UnitTest.swift (hunk 13 above)
Comment (mrehan27) [reply Tests/DataPipeline/UnitTest.swift]: Not exactly how it is written. We can probably use `CustomerIO.shared` too but let's stick to it for unit tests. Improved the comment.

File: Tests/DataPipeline/Support/XCTestCaseExtensions.swift (hunk 2 above)
Comment (levibostian) [reply Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: > I couldn't find a similar method in the XCTest framework. It may be likely due to Any not being Equatable, making direct comparison of [String : Any] challenging

Ah, that could be. I can see value in a convenient [String: Any] equal assertion function. 

If the goal is checking equality of 2 `[String: Any]` objects, I do think there could be a more simple and flexible solution. One that is also easier to maintain in the future. 

I checked some of the exiting custom attribute tests in the code base since those existing tests would also be checking `[String: Any]` equality. 

There might be another strategy being used in the code base today, but [what I see thus far](https://github.com/customerio/customerio-ios/blob/10d9f6cde37af5001160b0a9af853618ac3440e3/Tests/Tracking/CustomerIOImplementationTest.swift#L61) is to create a JSON string from the `[String: Any]` data type and then do a string comparison of the 2 strings. 

Do you think this could work for your use case?

File: Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (levibostian) [reply Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift]: > but then this can raise more challenges impacting other tests which currently I didn't plan for this PR and the ticket.

I do understand that a change to a base class `tearDown` could make other existing test functions fail. Meaning that a change like this could be scope creep as other tests will need to be changed. 

If a change like resetting the SDK state in `tearDown` does indeed make existing tests break, that is a problem. Test functions should not pass or fail, depending on the result of test functions that execute before it. The fact that test functions fail indicates we have bad tests that need addressed. 

From this point, I am OK with making this change in another PR as long as a ticket is created planning on making this change.

File: Tests/DataPipeline/DataPipelineInteractionTests.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (levibostian) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift:50]: I may have read this wrong. I thought the `guard` was checking that the array had a value at index 0. Instead, the `guard` is checking that the value at index 0 is not nil. 

The `guard` statement might be doing the same thing as `XCTAssertNotNil` in that case. 

I do find the `guard` to be boilerplate heavy which is where my comments come in. But, this is not a blocker.

File: Tests/DataPipeline/DataPipelineInteractionTests.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (levibostian) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift]: > And there are a lot of objects that Segment attaches to an event, and they are fine to have.

Good to know. With this in mind, my original suggestion would not work. 

I agree, I think the existing code does the job.

**Review by levibostian — APPROVED** [review]
Rest of suggestions non-blockers 👍

File: Tests/DataPipeline/Support/XCTestCaseExtensions.swift (hunk 2 above)
Comment (mrehan27) [reply Tests/DataPipeline/Support/XCTestCaseExtensions.swift:6]: Unfortunately this seems to work well for `[String: String]` only and not `[String: Any]`

Merging this for now, but keep sharing ideas so we can improve it and opt for simpler solution if possible.

File: Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineImplementation+ScreenViewsTest.swift]: Agreed, created [ticket](https://linear.app/customerio/issue/MOB-50/base-unittest-and-integrationtest-classes) to tackle this

File: Tests/DataPipeline/DataPipelineInteractionTests.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrehan27) [reply Tests/DataPipeline/DataPipelineInteractionTests.swift:50]: Sorry, I should have explained better with code earlier. Imagine a case where we get no events in `outputReader.identifyEvents` due to some bug and the `outputReader.identifyEvents.count` is `0`. Now let's execute code line by line

**Option 1**

```
// assertion fails but code doesn't stop here
XCTAssertEqual(outputReader.identifyEvents.count, 1)
// code crashes because last element is nil
let identifyEvent = outputReader.identifyEvents.last!
//
XCTAssertEqual(identifyEvent.userId, givenIdentifier)
```

**Option 2**

```
// assertion fails but code doesn't stop here
XCTAssertEqual(outputReader.identifyEvents.count, 1)
// identifyEvent is nil
let identifyEvent = outputReader.identifyEvents.last
// assertion fails but code doesn't stop here
XCTAssertNotNil(identifyEvent)
// code crashes because identifyEvent is nil here
XCTAssertEqual(identifyEvent!.userId, givenIdentifier)
```

**Only workable option**

```
// assertion fails
XCTAssertEqual(outputReader.identifyEvents.count, 1)
// identifyEvent is nil
let identifyEvent = outputReader.identifyEvents.last
// assertion fails
XCTAssertNotNil(identifyEvent)
// every assertion below this will surely fail
XCTAssertEqual(identifyEvent?.userId, givenIdentifier)
```

I preferred using `guard` statement here because it kept the code cleaner. If the event is `nil`, any line after `XCTAssertNotNil` will surely fail, which is why `guard` makes more sense to me in this context. Also, it helps avoid the complexity of optional chaining, which can lead to nested calls becoming cluttered with extra parentheses and `?` operators.
